Add PI support for more operations

RyanJHeld opened the pull request:
I got to have some fun with `atto` parsers in this one. I added a few parsers to allow parsing expressions that have lists for arguments (and even nested lists). I also added parsers that return `latis.data.Data` objects. I needed that to support `fromArgs` in the `Evaluation` operation, but if we decide we don't want those parsers we could always change the constructor for `Evaluation` to take something other than `Data`.

I moved the creation of operations from FdmlReader to UnaryOperation
Added support for the following operations in processing instructions:
- curry
- pivot
- evaluation

Commit:
Add PI support for more operations

Move the creation of operations from FdmlReader to UnaryOperation
Add support for the parsing of operations:
- curry
- pivot
- evaluation

lindholc commented on core/src/main/scala/latis/ops/Curry.scala:
```diff
@@ -67,3 +70,15 @@ case class Curry(arity: Int = 1) extends GroupOperation {
+object Curry {
+
+  def fromArgs(args: List[String]): Option[UnaryOperation] = args match {
+    case arity :: Nil => Try(arity.toInt) match {
```
I think it would be better to use `Either.catchNonFatal` or `Either.catchOnly[NumberFormatException]`.

lindholc commented on core/src/main/scala/latis/ops/Curry.scala:
```diff
@@ -67,3 +70,15 @@ case class Curry(arity: Int = 1) extends GroupOperation {
+
+object Curry {
+
+  def fromArgs(args: List[String]): Option[UnaryOperation] = args match {
```
I imagine we'd want this to return `Option[Curry]`? Or `Either[LatisException, Curry]`?

RyanJHeld replied:
About returning `Curry`, will do.
About returning `Either`, I chose `Option` with the thought that someday someone might want to provide their own operation and need to define their own `fromArgs`, and we don't want other people throwing `LatisException` around. But we agreed this was far enough out that I can just use `Either` here if you'd like. Thoughts?

dlindhol replied:
The value of a useful error message seems to trump that.

I'm imagining a layer (even in Java, python, ...) that 3rd-party developers could implement. We could wrap it and any exceptions it might throw, on purpose or otherwise.

lindholc replied:
I think returning a `LatisException` is fine for now. When we actually support third-party operations we might choose to replace it with a different error type.

RyanJHeld commented on core/src/main/scala/latis/ops/parser/parsers.scala:
```diff
@@ -115,4 +127,38 @@ object parsers {
+  } yield Data.BooleanValue(n.toBoolean)
+
+  def floatValue: Parser[Data] = for {
+    n <- decimal
```
```suggestion
    n <- scientific | decimal
```

Commit:
Fix floatValue parser

A float can be specified in scientific notation, so look for that before decimal.

RyanJHeld commented on core/src/main/scala/latis/ops/parser/parsers.scala:
```diff
@@ -115,4 +127,38 @@ object parsers {
+    _ <- char('f') | char('F')
+  } yield Data.FloatValue(n.toFloat)
+
+  def doubleValue: Parser[Data] = for {
```
I'm just realizing a problem with this parser. You can specify a double by giving an int followed by `d` or `D`. The `floatValue` parser could be fixed by just adding `| integer` in the for comprehension, but this one will take a little more thinking since we want scientific and decimal numbers to default to double if there is no `f` or `d` specified.

dlindhol replied:
It seems like taking on Data parsing is tricky business when we don't know what type we need until we have the scalar definition to tell us how to interpret it.

How does scala parse the literal "1d"? If we want to go here, I'd rather delegate the parsing then use `Data.fromValue`.

lindholc commented on core/src/main/scala/latis/input/fdml/FdmlReader.scala:
```diff
@@ -79,21 +79,17 @@ object FdmlReader {
-      case ast.Operation("project", vs) => Right(ops.Projection(vs: _*))
-      case ast.Operation(op, _) => Left(LatisException(s"Unknown operation: $op"))
+      case ast.Operation(name, args) =>
+        UnaryOperation.makeOperation(name, args) match {
```
Take a look at the `toRight` method.

RyanJHeld replied:
I gotta get better at spotting those. I think I tried `fold`, but the compiler couldn't figure out what types go in the `Either`.

lindholc commented on core/src/main/scala/latis/ops/UnaryOperation.scala:
```diff
@@ -19,3 +19,15 @@ trait UnaryOperation extends Operation {
+
+object UnaryOperation {
+
+  def makeOperation(name: String, args: List[String]): Option[UnaryOperation] = name match {
```
I think it would be good to fail with a reason, like the operation name wasn't recognized or that the arguments for the operation couldn't be parsed.

lindholc commented on core/src/test/scala/latis/ops/UnaryOperationSpec.scala:
```diff
@@ -0,0 +1,25 @@
+    makeOperation("evaluation", List("1")) should be (Some(Evaluation(Data.IntValue(1))))
+    inside(makeOperation("evaluation", List("1", "b"))) { case Some(Evaluation(t)) =>
+      t.asInstanceOf[TupleData].elements should be (Seq(Data.IntValue(1), Data.StringValue("b")))
+    }
```
Is there a reason for casting instead of pattern matching a `TupleData`?

RyanJHeld replied:
I can't get a pattern match to work. I tried `case Some(Evaluation(TupleData(es)))` to get `elements` out, but it didn't match. I just tried `Some(Evaluation(t@TupleData(_)))`, but still nothing. It says it couldn't match `Some(Evaluation(latis.data.TupleData@30534c5e))`.

dlindhol replied:
How about `case t: TupleData`?
We do have `TupleData.unapplySeq` so I'm not sure why that doesn't work.

lindholc replied:
I think this ought to work:

```scala
case Some(Evaluation(t: TupleData)) =>
```

You need something like this to use `unapplySeq`:

```scala
case Some(Evaluation(TupleData(elems@_*)) =>
```

lindholc commented on core/src/test/scala/latis/ops/parser/ParsersSpec.scala:
```diff
@@ -0,0 +1,43 @@
+class ParsersSpec extends FlatSpec {
+
+  "parsers.data" should "parse to Data of the correct type" in {
+    def testData = testParser(parsers.data)(_, _)
```
This (and `testOperation`) are probably better defined as `val`s so they are evaluated once rather than once per call.

lindholc commented on core/src/test/scala/latis/ops/parser/ParsersSpec.scala:
```diff
@@ -0,0 +1,43 @@
+    case ParseResult.Fail(_, _, m) => throw new Exception(s"$m in $s")
+    // parseOnly will never return anything but Done or Fail, but the types don't
+    // know that so we get a warning without the following line.
+    case _ => throw new Exception(s"failed to parse $s")
```
Is there a reason for throwing exceptions rather than `fail`ing?

RyanJHeld replied:
No reason. I'll change it.

dlindhol commented on core/src/main/scala/latis/ops/Evaluation.scala:
```diff
@@ -29,3 +33,11 @@ case class Evaluation(data: Data) extends UnaryOperation {
+  def fromArgs(args: List[String]): Option[UnaryOperation] = for {
+    ds <- args.traverse(data.parseOnly(_).option)
+    d = Data.fromSeq(ds)
+  } yield Evaluation(d)
```
I think this will be a case like `pivot` where we won't know what value type we need until it is  invoked with the `DataType` (i.e. model). Sounds like we need to settle for a string argument (although it is a red flag that we should be able to design `Operations` better).

RyanJHeld replied:
That's understandable. I'll make the change and delete the `Data` parsers.

dlindhol commented on core/src/main/scala/latis/ops/Projection.scala:
```diff
@@ -60,4 +60,9 @@ object Projection {
 
   def apply(exp: String): Projection =
     Projection(exp.split(",").toIndexedSeq: _*)
```
This could delegate to the new `fromArgs`. 

I generally like the pattern of having a single definitive "smart" constructor that can call `new`. Other convenient constructor methods would delegate to it to minimize repetitive logic. This case with a public case class and nothing "smart" to do might not call for it.

RyanJHeld replied:
That means we would be changing the type to return `Option` or `Either`, and I know you don't like having `apply` functions return something other than the class. Which way should we go here?

dlindhol replied:
I guess that really argues for getting rid of the `apply` and making the constructor private. We have a lot of that need going around. Sounds like a task for another day (#134 ).

dlindhol commented on core/src/main/scala/latis/ops/Rename.scala:
```diff
@@ -21,3 +21,11 @@ case class Rename(origName: String, newName: String) extends UnaryOperation {
+
+  def fromArgs(args: List[String]): Option[UnaryOperation] = args match {
+    case oldName :: newName :: Nil => Some(Rename(oldName, newName))
+    case _ => None
```
I could imagine some useful error message (and thus `Either`) here.

Commit:
Fix things for PR

deleted `Data` parsers
Evaluation takes `String` as input instead of `Data`
Operation.fromArgs constructors return `Either` instead of `Option`

RyanJHeld commented on core/src/main/scala/latis/ops/parser/parsers.scala:
```diff
@@ -3,7 +3,7 @@ package latis.ops.parser
 import atto.Parser
 
-import ast._
+import latis.ops.parser.ast._
```
IntelliJ must have done this automatically...

RyanJHeld commented on core/src/test/scala/latis/ops/EvaluationSpec.scala:
```diff
@@ -12,29 +12,12 @@ class EvaluationSpec extends FlatSpec {
     }
   }
 
-  "Evaluation" should "evaluate a 2D dataset" in {
-    val d = TupleData(1, 4)
-    DatasetGenerator.generate2DDataset(
-      Vector(0, 1, 2),
-      Vector(3, 4),
-      Vector(
-        Vector(10, 20),
-        Vector(12, 22),
-        Vector(14, 24)
-      )
-    ).withOperation(Evaluation(d)).unsafeForce().data match {
-      case ConstantFunction(Number(d)) =>
-        d should be (22)
-    }
-  }
-
-
```
This is a sad consequence of changing the constructor type to `String`. We could do some parsing in `Evaluation.applyToData` and try to make a `TupleData` from a string like `(1, 4)`.

RyanJHeld commented on core/src/main/scala/latis/ops/UnaryOperation.scala:
```diff
@@ -19,3 +20,19 @@ trait UnaryOperation extends Operation {
+    case "curry" => Curry.fromArgs(args)
+    case "pivot" => Pivot.fromArgs(args)
+    case "evaluation" => Evaluation.fromArgs(args)
+    case "eval" => Evaluation.fromArgs(args)
```
I added this `eval` alias because I noticed the same alias defined in `datasets.package.scala`. I want to point it out to make sure everyone is okay with it.

dlindhol commented on core/src/main/scala/latis/dataset/package.scala:
```diff
@@ -21,7 +21,7 @@ package object dataset {
     def contains(varName: String, values: String*): Dataset = dataset.withOperation(Contains(varName, values: _*))
     def rename(varName: String, newName: String): Dataset = dataset.withOperation(Rename(varName, newName))
-    def eval(value: Data): Dataset = dataset.withOperation(Evaluation(value))
+    def eval(value: String): Dataset = dataset.withOperation(Evaluation(value))
```
From the DSL perspective I'm tempted to support `Any` here. For example `myDataset.eval(1.2)`. We could do a `toString` for starters.

It just struck me that we need to be able to evaluate with a tuple. 'myDataset.eval(1.2, "foo")'? Maybe `Evaluation(value: String*)`? We can make a new ticket for that.
I'm also tempted to use `apply` here so we can really treat our datasets like functions:  'myDataset(1.2, "foo")'.

It keeps bugging me that we often want to construct an Operation that is agnostic to the Dataset it will be applied to but needs to know the model of the Dataset to really be the type safe thing we want. It does seem to call for an abstraction that is not yet an Operation until it can be attached to a DataType. Something akin to the ast.CExpr? Sounds like an off-sprint discussion.

sapols commented on core/src/test/scala/latis/ops/parser/ParsersSpec.scala:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  /**
+   * partially apply with a parser to get a function that takes the string you
```
It bothers me the tiniest bit that "partially" isn't capitalized. 🤷‍♂️

RyanJHeld replied:
```suggestion
   * Partially apply with a parser to get a function that takes the string you
```

Commit:
Update core/src/test/scala/latis/ops/parser/ParsersSpec.scala
